File: benchmarks/C++/src/json/ParseException.cpp
#include "ParseException.h"

namespace json {
    ParseException::ParseException(string message, int offset, int line, int column) {
        _offset = offset;
        _line   = line;
        _column = column;
        _what = message + " at " + to_string(line) + ":" + to_string(column);
    }
    
    int ParseException::getOffset() {
        return _offset;
    }

    int ParseException::getLine() {
        return _line;
    }

    int ParseException::getColumn() {
        return _column;
    }

    const char *ParseException::what() const throw() {
        return (_what.c_str());
    }

}

File: benchmarks/Rust/src/json/parse_exception.rs
use std::error::Error;
use std::fmt::Display;

#[derive(Debug)]
pub struct ParseException {
    message: String,
    offset: isize,
    line: usize,
    column: usize,
}

impl ParseException {
    pub fn new(message: String, offset: isize, line: usize, column: usize) -> ParseException {
        ParseException {
            message,
            offset,
            line,
            column,
        }
    }

    pub fn get_offset(&self) -> isize {
        self.offset
    }

    pub fn get_line(&self) -> usize {
        self.line
    }

    pub fn get_column(&self) -> usize {
        self.column
    }
}

impl Display for ParseException {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        let Self {
            message,
            line,
            column,
            ..
        } = self;
        write!(f, "{message} at {line}:{column}")
    }
}

impl Error for ParseException {}
